File: core/client/game/src/shine/support/pool/StringBuilderPool.cs
using System;
using System.Text;
using System.Threading;

namespace ShineEngine
{
	/// <summary>
	/// stringBuilder池
	/// </summary>
	public class StringBuilderPool:ObjectPool<StringBuilder>
	{
		private static StringBuilderPool _instance=new StringBuilderPool();

		private static SMap<Thread,StringBuilderPool> _dic=new SMap<Thread,StringBuilderPool>();

		public StringBuilderPool():base(()=>new StringBuilder())
		{

		}

		public override void back(StringBuilder obj)
		{
			if(obj.Length>ShineSetting.stringBuilderPoolKeepSize)
				return;

			obj.Length=0;

			base.back(obj);
		}

		private static StringBuilderPool getPoolByThread(Thread thread)
		{
			StringBuilderPool pool=_dic.get(thread);

			if(pool==null)
			{
				_dic.put(thread,pool=new StringBuilderPool());
			}

			return pool;
		}

		/** 取一个 */
		public static StringBuilder create()
		{
			return _instance.getOne();
		}

		/** 还一个 */
		public static void release(StringBuilder obj)
		{
			_instance.back(obj);
		}

		/** 还一个并返回该StringBuilder的字符串 */
		public static string releaseStr(StringBuilder obj)
		{
			string v=obj.ToString();

			_instance.back(obj);

			return v;
		}

		/** 取一个 */
		public static StringBuilder createForThread()
		{
			return getPoolByThread(Thread.CurrentThread).getOne();
		}

		/** 还一个 */
		public static void releaseForThread(StringBuilder obj)
		{
			getPoolByThread(Thread.CurrentThread).back(obj);
		}

		/** 还一个并返回该StringBuilder的字符串 */
		public static string releaseStrForThread(StringBuilder obj)
		{
			string v=obj.ToString();

			getPoolByThread(Thread.CurrentThread).back(obj);

			return v;
		}
	}
}

File: core/server/project/shine/src/main/java/com/home/shine/support/pool/StringBuilderPool.java
package com.home.shine.support.pool;

import com.home.shine.control.ThreadControl;
import com.home.shine.global.ShineSetting;
import com.home.shine.thread.AbstractThread;

/** stringBuilder池 */
public class StringBuilderPool extends ObjectPool<StringBuilder>
{
	/** 回车 */
	public static final String Enter="\r\n";
	/** 换行 */
	public static final String Tab="\t";
	
	public StringBuilderPool()
	{
		super(StringBuilder::new);
	}
	
	@Override
	public void back(StringBuilder obj)
	{
		//超出的不回收了
		if(obj.length()>= ShineSetting.stringBuilderPoolKeepSize)
			return;
		
		obj.setLength(0);
		
		super.back(obj);
	}
	
	/** 取一个 */
	public static StringBuilder create()
	{
		AbstractThread thread=ThreadControl.getCurrentShineThread();
		
		if(thread!=null)
		{
			return thread.stringBuilderPool.getOne();
		}
		else
		{
			return new StringBuilder();
		}
	}
	
	/** 取一个 */
	public static StringBuilder create(String str)
	{
		StringBuilder sb=create();
		sb.append(str);
		return sb;
	}
	
	/** 还一个 */
	public static void release(StringBuilder obj)
	{
		AbstractThread thread=ThreadControl.getCurrentShineThread();
		
		if(thread!=null)
		{
			thread.stringBuilderPool.back(obj);
		}
	}
	
	/** 还一个并返回该StringBuilder的字符串 */
	public static String releaseStr(StringBuilder obj)
	{
		String v=obj.toString();

		release(obj);

		return v;
	}
}
